connectors/services/consume.py

```diff
@@ -0,0 +1,152 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+"""
+Event loop
+
+- polls for work by calling Elasticsearch on a regular basis
+- instanciates connector plugins
+- mirrors an Elasticsearch index with a collection of documents
+"""
+import asyncio
+import os
+import time
+
+from connectors.byoc import BYOIndex, DataSourceError, PipelineSettings, SyncJobIndex
+from connectors.byoei import ElasticServer
+from connectors.es import DEFAULT_LANGUAGE, defaults_for
+from connectors.logger import logger
+from connectors.services.base import BaseService
+from connectors.source import DataSourceConfiguration, get_source_klass
+from connectors.utils import CancellableSleeps, trace_mem
+
+
+class JobService(BaseService):
+    def __init__(self, args):
+        super().__init__(args)
+        self.running = False
+
+        self._sleeps = CancellableSleeps()
+
+        elastic_config = self.config["elasticsearch"]
+
+        self.elastic_server = ElasticServer(elastic_config)
+        self.jobs = SyncJobIndex(elastic_config)
+        self.connectors = BYOIndex(elastic_config)
+
+        self.language_code = self.config.get(
+            "language_code", DEFAULT_LANGUAGE
+        )  # XXX: get language code from the connector instead
+        self.bulk_options = elastic_config.get("bulk", {})
+
+    async def run(self):
+        one_sync = self.args.one_sync
+        self.running = True
+
+        try:
+            while self.running:
+                if "connector_id" in self.config:
+                    connectors_ids = [self.config.get("connector_id")]
+                else:
+                    connectors_ids = []
+
+                query = self.jobs.build_docs_query(connectors_ids)
+
+                synced_anything = False
+                async for job in self.jobs.get_all_docs(query=query):
+                    await self.execute(job)
+                    synced_anything = True
+
+                if one_sync and synced_anything:
+                    logger.info("Ran a round of syncs and exiting.")
+                    break
+
+                await self._sleeps.sleep(10)
+        finally:
+            self.stop()
+        return 0
+
+    async def execute(self, job):
```

Comment:
Should claim/start the job before execution.

Reply:
True, added.

I found that current clients do not support optimistic concurrency and can cause race conditions - something we need to address too (can be done later).